clarity/utils/file_io.py

```diff
@@ -14,3 +25,131 @@ def write_jsonl(filename: str, records: list) -> None:
+        strict (bool): raise error if signal out of range for int16 (default: False).
+    """
+
+    if floating_point is False:
+        subtype = "PCM_16"
+        # Signal is float and we want to convert to int16
+        # *NB* Not  *= in next line as we need to make a copy
+        signal = signal * 32768
+        if strict and (np.max(signal) > 32767 or np.min(signal) < -32768):
+            raise ValueError("Signal out of range -1.0 to 1.0")
+        signal = signal.astype(np.dtype("int16"))
+    else:
+        subtype = "FLOAT"
+
+    soundfile.write(filename, signal, sample_rate, subtype=subtype)
+
+
+def read_signal(
+    filename: str | Path,
+    sample_rate: float = 0.0,
+    offset: int | float = 0,  # offset can be in samples or seconds
+    n_samples: int = -1,
+    n_channels: int = 0,
+    offset_is_samples: bool = False,
+    allow_resample: bool = True,
+) -> ndarray:
+    """Read a wav format audio file and return as numpy array of floats.
+
+    The returned value will be a numpy array of floats of shape (n_samples,
+    n_channels) for n_channels >= 2, and shape (n_samples,) for n_channels = 1.
+
+    If n_samples is set to a value other than -1, the specified number of samples
+    will be read, or until the end of the file. An 'offset' can be set to start
+    reading from a specified sample or time (in seconds).
+
+    The expected number of channels can be specified. If the file has a different
+    number of channels, an error will be raised.
+
+    The expected sample rate can be specified. If the file has a different sample
+    the file will be resampled to the expected sample rate, unless
+    'allow_resample' is set to False, in which case an error will be raised.
+
+    Args:
+        filename (str|Path): Name of file to read
+        sample_rate (float): The expected sample rate (default: 0.0 = any rate OK)
+        offset (int, optional): Offset in samples or seconds (from start).
+            Default is 0.
+        n_samples (int): Number of samples.
+        n_channels (int): expected number of channel (default: 0 = any number OK)
+        offset_is_samples (bool): is offset measured in samples, (True) or
+            seconds (False) (default: False)
+        allow_resample (bool): allow resampling if sample rate is different
+            from expected rate. Else raise error (default: True)
+
+    Returns:
+        np.ndarray: audio signal
+    """
+
+    wave_file = SoundFile(filename)
+
+    if n_channels not in (0, wave_file.channels):
+        raise ValueError(
+            f"Wav file ({filename}) was expected to have {n_channels} channels."
+        )
+
+    if not offset_is_samples:  # Default behaviour
+        offset = np.rint(offset * wave_file.samplerate).astype(int)
+
+    if offset != 0:
+        wave_file.seek(offset)
+
+    signal = wave_file.read(frames=n_samples)
+
+    if sample_rate not in (0, wave_file.samplerate):
+        if allow_resample:
```

Comment:
I see in `msbg_utils.read_signal_x()` that resampling is done and wonder if having a separate function for resampling might be more useful rather than having it embedded within the `read_signal()` function.

It would allow for a specific unit-test to be written and resampling to be performed anywhere within the code base.

More appropriate as separate issue though I think (if its a useful thing to do).

Reply:
Yes,  I think I will introduce a separate resample - that takes two sample rates , i.e. just a change to the scipy.resample semantics  -- but then call this from within read_signal().  I'd like to keep it embedded because it's such a common thing to do ... many of our algorithms are designed for specific sample rates. Also the librosa library -- very popular in the community -- has a similar semantic, so it won't seem so odd to people.